### DISCUSSION TITLE
character sets

### DISCUSSION THREAD

**Discussion by hammondm** [coqui-ai/TTS#1438] [General Q&A]
Hi

I've snooped around github, gitter, and the web for an answer to this, but I haven't found any yet.

I'm trying to train a tacotron2 model for Welsh. I've basically converted a subset of the Welsh Common Voice data into a form that looks like LJSpeech and then I use the LJSpeech recipe with little tweaks: file locations, no phonemes.

The problem is that it's telling me that characters are not found.

I gather I can set the characterset in a config.json file, but the recipe doesn't use a config.json file.

So my question is what is the syntax for setting the character set if I want to insert it directly into my tweaked recipe file?

thank you,

mike hammond

**Comment by hammondm** [comment]
Oh, I may have found an example in the multilingual VITS recipe.
:)

**Reply by hammondm** [reply]
No this did not work. This is the error:

```
(most recent call last):
  File "twone2.py", line 88, in <module>
    tokenizer, config = TTSTokenizer.init_from_config(config)
  File "/content/TTS/TTS/tts/utils/text/tokenizer.py", line 164, in init_from_config
    if config.characters and config.characters.characters_class:
AttributeError: 'dict' object has no attribute 'characters_class'
```

**Comment by arif334** [comment]
You can define a `CharactersConfig` object in your training script. Things you have to do are these:

* Import `CharactersConfig`:
```
from TTS.tts.configs.shared_configs import CharactersConfig
```

* Define your own `CharactersConfig` object. Here's an example:
```
characters = CharactersConfig(
    pad="_",
    eos="~",
    bos="^",
    characters="@ABCDEFGHIJKLMNPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz\u0981\u0982\u0983\u0984\u0985\u0986\u0987\u0988\u0989\u098a\u098b\u098c\u098d\u098e\u098f\u0990\u0991\u0992\u0993\u0994\u0995\u0996\u0997\u0998\u0999\u099a\u099b\u099c\u099d\u099e\u099f\u09a0\u09a1\u09a2\u09a3\u09a4\u09a5\u09a6\u09a7\u09a8\u09a9\u09aa\u09ab\u09ac\u09ad\u09ae\u09af\u09b0\u09b1\u09b2\u09b3\u09b4\u09b5\u09b6\u09b7\u09b8\u09b9\u09ba\u09bb\u09bc\u09bd\u09be\u09bf\u09c0\u09c1\u09c2\u09c3\u09c4\u09c5\u09c6\u09c7\u09c8\u09c9\u09ca\u09cb\u09cc\u09cd\u09ce\u09cf\u09d0\u09d1\u09d2\u09d3\u09d4\u09d5\u09d6\u09d7\u09d8\u09d9\u09da\u09db\u09dc\u09dd\u09de\u09df",
    punctuations=";:,.!?—…\"«»“”' ()[]/\u0964\u2019",
    phonemes="ɑɐɒæɓʙβɔɕçɗɖðʤəɘɚɛɜɝɞɟʄɡɠɢʛɦɧħɥʜɨɪʝɭɬɫɮʟɱɯɰŋɳɲɴøɵɸθœɶʘɹɺɾɻʀʁɽʂʃʈʧʉʊʋⱱʌɣɤʍχʎʏʑʐʒʔʡʕʢǀǁǂǃˈˌːˑʼʴʰʱʲʷˠˤ˞↓↑→↗↘'̩'ᵻ",
    unique=True
)
```
Note: If you are using TTS version 0.6.*, replace `unique` with `is_unique`

* Then add this object in the `Tacotron2Config`:
```
config = Tacotron2Config( 
...
characters=characters,
)
```

**Reply by hammondm** [reply]
Excellent, thank you. That seems to work.

I'm understanding that I do *not* have to specify the phonemes, right? That I can operate straight from orthography?

mike h.

**Reply by arif334** [reply]
> Excellent, thank you. That seems to work.
> 
> I'm understanding that I do _not_ have to specify the phonemes, right? That I can operate straight from orthography?
> 
> mike h.

Yes, `phonemes` are optional.

**Comment by hammondm** [comment]
Hi
Just a followup. This same technique doesn't seem to work for the VITS recipe. Is there some different move required there?
thanks,
mike h

**Reply by makerjackie** [reply]
use dev branch for now

**Reply by tieincred** [reply]
Hi @hammondm @Jackiexiao , I am on learning phase so please forgive for noob questions.
Adding to the problem said by @hammondm I am trying to train hindi data on vits, but when I add character set it seems to give me error.
RuntimeError: cuDNN error: CUDNN_STATUS_NOT_INITIALIZED
I have tried doing few things but that doesn't seem to be the solving the problem.
but without specifying characters it works
Is there anything that I have to change while specifying the characters/

**Reply by makerjackie** [reply]
> RuntimeError: cuDNN error: CUDNN_STATUS_NOT_INITIALIZED
check your gpu / cudnn

**Reply by tieincred** [reply]
@Jackiexiao, thanks for the reply.
I have checked it already, and training seems to going well for original LJspeech data set.
But when I train with character set it starts giving problem
Please let me know if you still think the problem is in gpu/cudnn
